File: 4242-sum-of-gcd-of-formed-pairs/sum-of-gcd-of-formed-pairs.c
#include <stdlib.h>

int gcd(int a, int b) {
    while (b != 0) {
        int temp = b;
        b = a % b;
        a = temp;
    }
    return a;
}

int compare(const void *a, const void *b) {
    return (*(int *)a - *(int *)b);
}

long long gcdSum(int* nums, int numsSize) {
    int* prefixGCD = (int*)malloc(sizeof(int) * numsSize);

    long long sumGcd = 0;
    int max = nums[0];

    for (int i = 0; i < numsSize; i++) {
        if (nums[i] > max)
            max = nums[i];
        prefixGCD[i] = gcd(max, nums[i]);
    }

    qsort(prefixGCD, numsSize, sizeof(int), compare);

    int i = 0, j = numsSize - 1;
    while (i < j) {
        sumGcd += gcd(prefixGCD[i], prefixGCD[j]);
        i++;
        j--;
    }

    free(prefixGCD);
    return sumGcd;
}

File: DailyStreak/4242-sum-of-gcd-of-formed-pairs/sum-of-gcd-of-formed-pairs.java
class Solution {
    public long gcdSum(int[] nums) {
        int[] prefixGCD = new int[nums.length];
        long sumGcd = 0;
        int max = Integer.MIN_VALUE;

        for (int i = 0; i < nums.length; i++) {
            max = Math.max(max, nums[i]);
            prefixGCD[i] = getGcd(max, nums[i]);
        }

        Arrays.sort(prefixGCD);

        for (int i = 0, j = prefixGCD.length - 1; i < j; i++, j--) {
            sumGcd += getGcd(prefixGCD[i], prefixGCD[j]);
        }

        return sumGcd;
    }

    private int getGcd(int a, int b) {
        while (b != 0) {
            int temp = b;
            b = a % b;
            a = temp;
        }
        return a;
    }
}